packages/assets-controllers/src/RatesController/RatesController.ts

```diff
@@ -0,0 +1,197 @@
+import { BaseController } from '@metamask/base-controller';
+import { Mutex } from 'async-mutex';
+
+import { fetchMultiExchangeRate as defaultFetchExchangeRate } from '../crypto-compare-service';
+import type {
+  ConversionRates,
+  RatesState,
+  RatesMessenger,
+  RatesControllerArgs,
+} from './types';
+
+export const name = 'RatesController';
+
+const DEFAULT_CURRENCIES = {
+  btc: 'btc',
+};
+
+const metadata = {
+  currency: { persist: true, anonymous: true },
+  rates: { persist: true, anonymous: true },
+  cryptocurrencyList: { persist: true, anonymous: true },
+};
+
+const defaultState = {
+  currency: 'usd',
+  rates: {
+    btc: {
+      conversionDate: 0,
+      conversionRate: 0,
+      usdConversionRate: null,
+    },
+  },
+  cryptocurrencyList: [DEFAULT_CURRENCIES.btc],
+};
+
+export class RatesController extends BaseController<
+  typeof name,
+  RatesState,
+  RatesMessenger
+> {
+  readonly #mutex = new Mutex();
+
+  readonly #fetchMultiExchangeRate;
+
+  readonly #onStart;
+
+  readonly #onStop;
+
+  readonly #includeUsdRate;
+
+  #intervalId: NodeJS.Timeout | undefined;
+
+  #intervalLength: number | undefined;
+
+  /**
+   * Creates a BtcRateController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.includeUsdRate - Keep track of the USD rate in addition to the current currency rate.
+   * @param options.interval - The polling interval, in milliseconds.
+   * @param options.messenger - A reference to the messaging system.
+   * @param options.state - Initial state to set on this controller.
+   * @param options.fetchMultiExchangeRate - Fetches the exchange rate from an external API. This option is primarily meant for use in unit tests.
+   * @param options.onStart - Optional callback to be executed when the polling stops.
+   * @param options.onStop - Optional callback to be executed when the polling starts.
+   */
+  constructor({
+    interval = 180000,
+    messenger,
+    state,
+    includeUsdRate,
+    fetchMultiExchangeRate = defaultFetchExchangeRate,
```

Comment:
I have an ADR for the service pattern already: https://github.com/MetaMask/decisions/pull/17/files

Still in draft but only because I wanted to do a final editing pass, and add a diagram.

It doesn't include any mention of an abstract class, but I am not sure that pattern has any benefits. I was going to suggest we delete it, in the case of the token price services. I had suggested initially that we consider abstracting the implementation away so that we could toggle between the two token price endpoints, but we never actually needed to do that.

Reply:
Hmm, I think we'd need a type to represent the service so that we wouldn't _need_ to pass a CryptoCompareService to the RatesController in this case (we could pass a mock version in tests, for instance, that implements the same methods but with mock implementations). Maybe there are multiple ways to do this however.

Reply:
I agree there's an improvement to be done here. I would advocate for this conversation, decision, and later development to be done outside this PR.